### PR TITLE
Added limit and left for tasks

### PR DESCRIPTION
**Pull request by Bibitha-V**

### CODE REVIEW

File: web/engine/data.js [hunk 1]
```diff
@@ -221,10 +220,58 @@ More details should be available in the log file: ${log.getName()}
...
+    if (finalObj) {
+      for (let i = 0; i < finalObj.length; i++) {
+        let curr = finalObj[i].action;
+        if (!attsummary[curr]) {
+          attsummary[curr] = {
+            attempt: 0,
+          };
+        } else attsummary[curr].attempt++;
+      }
+    }
```
Comment (theproductiveprogrammer) [inline web/engine/data.js:265]: This could have been done earlier. The number of steps seems to be a bit too many

File: web/engine/data.js [hunk 2]
```diff
@@ -221,10 +220,58 @@ More details should be available in the log file: ${log.getName()}
...
+    
+  return attsummary;
+}
+
```
Comment (theproductiveprogrammer) [inline web/engine/data.js:269]: # Sample Rewrite FYI & Reference

Try using/rewriting like this

```javascript
/*    problem/
 * return a summary of the number of tasks performed, grouped
 * by task type (action)
 *
 *    understand/
 * The DB data is stored in a per-user, per-task format with each task update listed in it's "steps" field:
 *    {
 *      userid: {
 *        id1: {
 *          ...
 *          steps: {
 *            { t: "2020-11-02T05:01:13.592Z", e: "task/new", data: { action: "LINKEDIN_VIEW" ...} ...},
 *            { t: "....", e: "task/status" data: { msg: "task/done", code: 200 ...} ... },
 *            { t: "....", e: "task/status" data: { msg: "task/completed", code: 202 ...} ... },
 *            { ... task update ...},
 *            { ... task update ...},
 *          }
 *        },
 *        id2: ...,
 *        id3: ...,
 *      }
 *
 *    way/
 * Walk the steps and count those completed within the last 24
 * hours against that task action.
 */
function getTasksPerDay(userid) {
  const tasksDoneToday = {}

  const tasks = DB[userid]
  if(!tasks) return tasksDoneToday

  const currentTime = Date.now()
  for(const task in tasks) {
    let current_action = null
    task.steps.map(t => {
      if(t.data.action) current_action = t.data.action
      if(!tasksDoneToday[current_action]) tasksDoneToday[current_action] = 0
      if(t.data.code != 202) return
      const tt = (new Date(t.t)).getTime()
      const hours = Math.floor((currentTime - tt) / 3600000);
      if(hours < 24) tasksDoneToday[current_action] += 1
    })
  }
  return tasksDoneToday
}
```

Note that the function is much smaller and simpler. We would probably want to make it more resilient by adding error checks:

```javascript
function getTasksPerDay(userid) {
  const tasksDoneToday = {}

  const tasks = DB[userid]
  if(!tasks) return tasksDoneToday

  const currentTime = Date.now()
  for(const task in tasks) {
    let current_action = null
    task.steps.map(t => {
      if(!t.data) return
      if(t.data.action) current_action = t.data.action
      if(!current_action) return
      if(!tasksDoneToday[current_action]) tasksDoneToday[current_action] = 0
      if(t.data.code != 202) return
      if(!t.t) return
      const tt = (new Date(t.t)).getTime()
      const hours = Math.floor((currentTime - tt) / 3600000);
      if(hours < 24) tasksDoneToday[current_action] += 1
    })
  }
  return tasksDoneToday
}
```

**Review by theproductiveprogrammer — CHANGES_REQUESTED** [review]

File: web/engine/data.js [hunk 3]
```diff
@@ -221,10 +220,58 @@ More details should be available in the log file: ${log.getName()}
...
 }
 
+
+function getTasksPerDay(userid) {
+  let userTaskArr=DB[userid]
+      let tasks = [];
+      for (const task in userTaskArr) {
+        tasks.push(userTaskArr[task]);
+      }
+    let [...taskArr] = tasks;
```
[the remark this answers is not in the document]
Comment (Bibitha-V) [reply web/engine/data.js:230]: I was trying the array approach for looping

File: web/engine/data.js [hunk 4]
```diff
@@ -221,10 +220,58 @@ More details should be available in the log file: ${log.getName()}
...
+    let taskType = taskArr.map((task) => {
```
[the remark this answers is not in the document]
Comment (Bibitha-V) [reply web/engine/data.js:231]: incorrect naming convention

File: web/engine/data.js [hunk 5]
```diff
@@ -221,10 +220,58 @@ More details should be available in the log file: ${log.getName()}
...
+      let list = task.steps.map((el) => el.data);
+      return list;
+    });
+
+    let finalObj = taskType.map((el) => {
```
[the remark this answers is not in the document]
Comment (Bibitha-V) [reply web/engine/data.js:236]: attempt summary

File: web/engine/data.js (hunk 1 above)
Comment (Bibitha-V) [reply web/engine/data.js:265]: updated as per the review comments

**Review by theproductiveprogrammer — APPROVED** [review]

**Pull request merged**